performance.md (#173)
Azure/azure-webpubsub: docs/concepts/performance.md

This guide can't cover all scenarios (and different use cases, message sizes, message sending patterns, and so on). But it provides some basic information to understand the performance limitation.

## Performance insight

This section describes the performance evaluation methodologies, and then lists all factors that affect performance. In the end, it provides methods to help you evaluate performance requirements.

### Methodology

*Throughput* and *latency* are two typical aspects of performance checking. For Azure Web PubSub Service, each SKU tier has its own throughput throttling policy. The policy defines *the maximum allowed throughput (inbound and outbound bandwidth)* as the maximum achieved throughput when 99 percent of messages have latency that's less than 1 second.

### Performance factors

Theoretically, Azure Web PubSub Service capacity is limited by computation resources: CPU, memory, and network. For example, more connections to Azure Web PubSub Service cause the service to use more memory. For larger message traffic (for example, every message is larger than 2,048 bytes), Azure Web PubSub Service needs to spend more CPU cycles to process traffic.

> The message routing cost also limits performance. Azure Web PubSub Service plays a role as a message router, which routes the message among a set of clients. A different scenario or API requires a different routing policy. 

Reviewer: message router => message broker?

> For **echo**, the client sends a message to itself, and the routing destination is also itself. This pattern has the lowest routing cost. But for **broadcast**, **send to group**, and **send to connection**, Azure Web PubSub Service needs to look up the target connections through the internal distributed data structure. This extra processing uses more CPU, memory, and network bandwidth. As a result, performance is slower.

Reviewer: > sends a message to itself

feels a little confusing, a client is unable to send a message to itself. How about: "sends a message to the upstream, and upstream echos the message back to the client"?

In summary, the following factors affect the inbound and outbound capacity:

-   SKU tier (CPU/memory)

-   Number of connections

-   Message size

-   Message send rate

> - Use-case scenario (routing cost)

Reviewer: nit: align with above items
```suggestion
-   Use-case scenario (routing cost)
```

### Finding a proper SKU

How can you evaluate the inbound/outbound capacity or find which tier is suitable for a specific use case?

Every tier has its own maximum inbound bandwidth and outbound bandwidth. A smooth user experience is not guaranteed after the inbound or outbound traffic exceeds the limit. 

```
  inboundBandwidth = inboundConnections * messageSize / sendInterval
  outboundBandwidth = outboundConnections * messageSize / sendInterval
```

- *inboundConnections*: The number of connections sending the message.
- *outboundConnections*: The number of connections receiving the message.
- *messageSize*: The size of a single message (average value). A small message that's less than 1,024 bytes has a performance impact that's similar to a 1,024-byte message.

> - *sendInterval*: The time of sending one message. Typically it's 1 second per message, which means sending one message every second. A smaller interval means sending more message in a time period. For example, 0.5 seconds per message means sending two messages every second.

Reviewer: ```suggestion
- *sendInterval*: The interval for sending messages. For example, 1 second means sending one message every second. A smaller interval means sending more messages in a time period. For example, 0.5 seconds means sending two messages every second.
```

> - *Connections*: The committed maximum threshold for Azure Web PubSub Service for every tier. If the connection number is increased further, it will suffer from connection throttling.

Reviewer: ```suggestion
- *Connections*: The committed maximum threshold for Azure Web PubSub Service for every tier. Connections that exceed the threshold get throttled.
```

> Assume that the upstream is powerful enough and is not the performance bottleneck. Then, check the maximum inbound and outbound bandwidth for every tier.

Reviewer: > Then, check the maximum inbound and outbound bandwidth for every tier.

What does this mean? Is it to "use the maximum inbound and outbound bandwidth to choose the proper SKU"?

## Case study

The following sections go through 3 typical use cases: **send to groups through Web PubSub subprotocol**, **triggering CloudEvent**, **calling rest api**. For each scenario, the section lists the current inbound and outbound capacity for Azure Web PubSub Service. It also explains the main factors that affect performance.

In all use cases, the default message size is 2,048 bytes, and the message send interval is 1 second.

### Send to groups through Web PubSub subprotocol

> The service supports a specific subprotocol called json.webpubsub.azure.v1 which empowers the clients to do publish/subscribe directly instead of a round trip to the upstream server. This scenario is quite efficient as no server is involved and all traffic goes through websocket.

Reviewer: ```suggestion
The service supports a specific subprotocol called `json.webpubsub.azure.v1` which empowers the clients to do publish/subscribe directly instead of a round trip to the upstream server. This scenario is quite efficient as no server is involved and all traffic goes through the client-service WebSocket connection.
```

> <img src="../images/wps-group.png" alt="wps-group" style="zoom:60%;" />

Reviewer: nice graph 👍

The routing cost is significant for sending message to many small groups. Currently, the Azure Web PubSub Service implementation hits the routing cost limit at Unit 50. Adding more CPU and memory doesn't help, so Unit100 can't improve further by design. If you need more inbound bandwidth, contact customer support.

|   Send to small group     | Unit1 | Unit2 | Unit5  | Unit10 | Unit20  | Unit50 | Unit100 |
|---------------------------|-------|-------|--------|--------|--------|--------|---------|
| Connections               | 1,000 | 2,000 | 5,000  | 10,000 | 20,000 | 50,000 | 100,000 |
| Group member count        | 10    | 10    | 10     | 10     | 10     | 10     | 10 |
| Group count               | 100   | 200   | 500    | 1,000  | 2,000  | 5,000  | 10,000 |
| Inbound messages per second  | 400 | 800  | 2,000 | 4,000 | 8,000 | 15,000 | 15,000 |
| Inbound bandwidth  | 800 KBps | 1.6 MBps | 4 MBps    | 8 MBps    | 16 MBps   | 30 MBps  | 30 MBps   |
| Outbound messages per second | 4,000 | 8,000 | 20,000 | 40,000 | 80,000 | 150,000 | 150,000 |
| Outbound bandwidth | 8 MBps   | 16 MBps | 40 MBps   | 80 MBps    | 160 MBps  | 300 MBps | 300 MBps |

### Triggering Cloud Event 

> Service delivers client events to the upstream webhook using the [CloudEvents HTTP protocol](https://github.com/cloudevents/spec/blob/v1.0.1/http-protocol-binding.md).

Reviewer: Use the link of our CloudEvents doc instead 
```suggestion
Service delivers client events to the upstream webhook using the [CloudEvents HTTP protocol](../references/protocol-cloudevents.md).
```

<img src="../images/upstream.png" alt="upstream" style="zoom:60%;" />

For every event, it formulates an HTTP POST request to the registered upstream and expects an HTTP response. 

> [!NOTE]

> > Web Pub Sub also supports HTTP 2.0 for upstream events delivering. The below result is tested using HTTP 1.1 . If your app server supports HTTP 2.0, the performance will be better. 

Reviewer: ```suggestion
> Web PubSub also supports HTTP 2.0 for upstream events delivering. The below result is tested using HTTP 1.1. If your app server supports HTTP 2.0, the performance will be better. 
```

#### Echo

In this case, the app server writes back the original message back in the http response. The behavior of **echo** determines that the maximum inbound bandwidth is equal to the maximum outbound bandwidth. For details, see the following table.

|       Echo                        | Unit1 | Unit2 | Unit5 | Unit10 | Unit20 | Unit50 | Unit100 |
|-----------------------------------|-------|-------|-------|--------|--------|--------|---------|
| Connections                       | 1,000 | 2,000 | 5,000 | 10,000 | 20,000 | 50,000 | 100,000 |
| Inbound/outbound messages per second | 500 | 1,000 | 2,500 | 5,000 | 10,000 | 25,000 | 50,000 |
| Inbound/outbound bandwidth | 1 MBps  | 2 MBps  | 5 MBps | 10 MBps  | 20 MBps  | 50 MBps | 100 MBps  |

### Rest API

> Sending high-density messages through the REST API is not as efficient as using WebSocket. 

Reviewer: We don't have a `using WebSocket` scenario, so maybe instead of the comparison, explain a little bit when to use REST API and simply show the performance test result below.

Author: Is there a doc or swagger introducing our rest API?

Reviewer: https://docs.microsoft.com/en-us/rest/api/webpubsub/
